### ISSUE TITLE
Starting Kubernetes Observability through basic SLIs/SLOs

### ISSUE THREAD

**Issue by gizas** [elastic/integrations#4997]
## Context

Observability definition includes much more than collection of metrics and logs. We should be able to provide to our users a holistic approach where functionalities like alerting and SLO definition should be part of the user experience for Kubernetes observability. In more details, we should help our users, when they onboard a new Kubernetes Cluster, to configure specific **Alerts** and **SLOs** and link those with specific instances of the data collected.

## Action

As part of this story we should :
- List representative categories of SLIs/ SLOs for Kubernetes 
- List basic starting alarms on the categories that were defined in previous step
- Provide Config As Code samples for applying above suggestions 
- Open any new enhancement stories (Kibana enhancements or Actionable-Observability team in case basic functionality is missing) when the UX needs improvement. Main goal is to link Alerting and SLOs with data collected from Kubernetes Integration side

## Useful Links
SLO API: https://github.com/kdelemme/kibana/blob/main/x-pack/plugins/observability/dev_docs/slo.md#L0-L1
Alert API: https://www.elastic.co/guide/en/kibana/current/alerting-apis.html

## Deliverables
- Document categories of SLIs and SLOs for Kubernetes
- Configuration files for Alerts and SLOs
- Any feature stories for UX enhancement: (https://github.com/elastic/package-spec/issues/484)
- Demoing e2e experience and probably updating some relevant blogs for Kubernetes Observability through SLI/ SLOs

## Relevant useful links:
- https://www.elastic.co/guide/en/kibana/current/alerting-getting-started.html
- https://www.squadcast.com/blog/using-observability-tools-to-set-slos-for-kubernetes-applications
- https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos

## TODOs
1. Investigate and collect categories of SLIs and SLOs for Kubernetes ✅ 
2. Investigate what alerts can be created based on the currently collected fields
  a. collect which metrics need to be collected/improved ✅ 
  b. Create snippets to install these alerts through the API ✅ 
3. Document the usage of these alerts and find proper place for official documentation  ✅ 
   a. PR to add them in k8s package docs: https://github.com/elastic/integrations/pull/5364
   b. PR to add them in automation repo: https://github.com/elastic/k8s-integration-infra/pull/16

Extras
5. Investigate how alerts can be part of the package: ✅  Issue created -> https://github.com/elastic/package-spec/issues/484 
6. Demo/[blogpost](https://docs.google.com/document/d/1NdYaF_KpW2ennI5SEC2H33n3tBJYLf7R1Y87xxe9Xv8/edit#): https://github.com/elastic/blog/issues/1869
7. Any follow up documentation.

**Comment by ChrsMark** [comment]
One basic SLI/SLO should be on `Leadership Switch`. 

Kubernetes native components including [cluster-autoscaler](https://pkg.go.dev/k8s.io/autoscaler/cluster-autoscaler), [kube-controller-manager](https://pkg.go.dev/k8s.io/kubernetes/cmd/kube-controller-manager), and [kube-scheduler](https://pkg.go.dev/k8s.io/kubernetes/cmd/kube-scheduler) are using leader-with-lease in [client-go](https://pkg.go.dev/k8s.io/client-go/tools/leaderelection). 

What we want to monitor here is:

1. (SLI_a) Leaderless: when there is no leader (need to define the SLI which would mean a critical error)
2. (SLI_b) Time of leadership switch (need to define the SLI which would be a warning)

Being with no leader for a period of time is quite critical for a production Kubernetes cluster. Hence we need to define proper SLIs/SLOs based on these observations.

This information can be retrieved by the `kube-state-metrics` Service and look like the following:

```
# HELP kube_lease_owner Information about the Lease's owner.
# TYPE kube_lease_owner gauge
kube_lease_owner{lease="kind-control-plane",owner_kind="Node",owner_name="kind-control-plane"} 1
# HELP kube_lease_renew_time Kube lease renew time.
# TYPE kube_lease_renew_time gauge
kube_lease_renew_time{lease="kind-control-plane"} 1.676268601e+09
```

SLO_a: `kube_lease_owner` should not be equal to zero for more than 30 seconds. That should indicate a CRITICAL error.
SLO_b: `avg(kube_lease_renew_time)` should not be greater than `0.5s` for a period of last 10 mins. That should indicate a WARNING.


At the moment we don't have a specific metricset/data_stream that specifically collects this information from `kube-state-metrics`. However we can create PoC by leveraging the Prometheus collector package.
@gizas what do you think? Shall we create separate issue for implementing the metricset to collect from `kube-state-metrics` or we can just use the Prometheus one?
I would be ok with doing both: creating an issue for the metricset implementation and in the meanwhile experiment by using the Prometheus package.

**Comment by gizas** [comment]
Leader SLIs is something needed, I agree. We have seen it in sdh recenlty.

But before discussing for specific slis, I would suggest to try to cover the big categories of SLIs as described from google here https://sre.google/workbook/implementing-slos/
Lets make a **doc** that will try to suggest and specify all indicators for each category. This way we will be specific and try to cover different types.

Now to your question: 
- Lets try to create indicators for what we can support now initially
- For any other cases agree lets create tickets and mention them here. 
> or we can just use the Prometheus one?
This means that we will force users to install Prometheues Integration if they want to have alerts right? I dont like so much

**Comment by ChrsMark** [comment]
As a first step I will focus on identifying the high level categories/groups of SLIs/SLOs.

One good approach is to follow the [Four Golden Signals](https://landing.google.com/sre/sre-book/chapters/monitoring-distributed-systems/#xref_monitoring_golden-signals) which are used by Google SRE. They are Latency, Traffic, Errors and Saturation.

A more specific approach is the [The RED method ](https://thenewstack.io/monitoring-microservices-red-method/) developed by Tom Wilkie, who was an SRE at Google and used the Four Golden Signals. The RED method drops the saturation because it is used for more advanced cases and people remember better things that come in threes :-). However I would personally be in favor of adding the saturation/utilization aspect into the game. 

In summary I would propose the following groups of SLIs/SLOs:
Group_1: latency (of control plane, like apiserver, controllermanager etc)
Group_2: resource utilisation (on nodes how much cpu, memory etc is used)
Group_3: errors (errors on logs or events, or error count from components, network etc)


If we agree on this first step as a base, I will then go ahead with defining specific SLIs/SLOs per group based on the metrics/data that we currently collect for the k8s cluster. @gizas what do you think about this plan?

**Comment by gizas** [comment]
Yes of course Chris. I think we are aligned, that is why I had added the 3rd link above in the description , to guide the assignee of the story to group the slos and justify the choices.

> However I would personally be in favor of adding the saturation/utilization aspect into the game.

You suggest this and you propose 3 Groups :) I got confused

I would suggest ( See examples from here https://sysdig.com/blog/golden-signals-kubernetes/)
Group_4: saturation (CPU usage, Disk space, Memory usage, percentage of pods in Nodes etc)

Just reminder that configuration of those through API is part of story. So lets think how we should save this information.
Maybe in a new folder called alerts inside package?

**Comment by ChrsMark** [comment]
> Yes of course Chris. I think we are aligned, that is why I had added the 3rd link above in the description , to guide the assignee of the story to group the slos and justify the choices.
> 
> > However I would personally be in favor of adding the saturation/utilization aspect into the game.
> 
> You suggest this and you propose 3 Groups :) I got confused
> 
> I would suggest ( See examples from here https://sysdig.com/blog/golden-signals-kubernetes/) Group_4: saturation (CPU usage, Disk space, Memory usage, percentage of pods in Nodes etc)
> 

You see 3 groups because I grouped latency and traffic into one group. In any case, this is just a conceptual grouping and 3 or 4 doesn't really matter in the end.

> Just reminder that configuration of those through API is part of story. So lets think how we should save this information. Maybe in a new folder called alerts inside package? Yaml or Json to support the API calls? ([Script](https://github.com/gizas/automatek8s/blob/TestBranch/converter_yaml_json.py) This was used to trnasform to json in previous relevant api work )

Yeap, that should happen at some point, it's also included in the TODOs of the issue.

**Comment by ChrsMark** [comment]
Heads up on this. After struggling a bit with ES queries and aggregations I managed to create one first alert using the k8s API request duration. One major issue here is that this metric is coming as a histogram from `kube-state-metrics` (Promtheus metric) but we store it not using our ES types which makes it hard then to calculate percentiles or anything useful.

See how we store these metrics at https://github.com/elastic/beats/blob/b0bbd16ed0064a4ef480ed15c96d0b954f0748a7/metricbeat/module/kubernetes/apiserver/_meta/testdata/docs.plain-expected.json#L12. @gizas we had discussed about this in the past with @MichaelKatsoulis and we concluded that maybe we need to add `types` support for these metricsets which use the prometheus helper. In that case storing these fields as ES histograms would help us to query for percentiles and having easier ES queries constructions. In my case I had to use runtime fields to overcome this.

I took the idea from `Average Apiserver Request Latency per Resource [Metrics Kubernetes]` visualization from our k8s dashboard which calculates the latency in milliseconds as `(sum(kubernetes.apiserver.request.duration.us.sum)/ sum(kubernetes.apiserver.request.duration.us.count))/1000`.

That's the very first alert which I put here as an example. I will continue with adding more similar ones with the currently available metrics. I will be updating this comment with the upcoming queries/alerts.

<details>
  <summary>k8s API request duration alert</summary>

  ```
PUT _watcher/watch/d2bb005f-55b0-4279-b597-09c4308e5352
{
  "trigger": {
    "schedule": {
      "interval": "30m"
    }
  },
  "input": {
    "search": {
      "request": {
        "body": {
          "query": {
            "bool": {
              "must": [
                {
                  "range": {
                    "@timestamp": {
                      "gte": "now-10m",
                      "lte": "now",
                      "format": "strict_date_optional_time"
                    }
                  }
                },
                {
                  "bool": {
                    "must": [
                      {
                        "query_string": {
                          "query": "event.module:kubernetes AND metricset.name: apiserver AND NOT (kubernetes.apiserver.request.verb: WATCH or kubernetes.apiserver.request.verb: CONNECT)",
                          "analyze_wildcard": true
                        }
                      },
                      {
                        "exists": {
                          "field": "kubernetes.apiserver.request.duration"
                        }
                      }
                    ],
                    "filter": [],
                    "should": [],
                    "must_not": []
                  }
                }
              ],
              "filter": [],
              "should": [],
              "must_not": []
            }
          },
          "size": 0,
          "runtime_mappings": {
            "avg_duration": {
              "type": "double",
              "script": {
                "source": "emit(doc['kubernetes.apiserver.request.duration.us.sum'].value / doc['kubernetes.apiserver.request.duration.us.count'].value / 1000 )"
              }
            }
          },
          "aggs": {
            "avg_duration": {
              "avg": {
                "field": "avg_duration"
              }
            }
          }
        },
        "indices": [
          "*"
        ]
      }
    }
  },
  "condition": {
    "compare": {
      "ctx.payload.aggregations.avg_duration": {
        "gte": 10
      }
    }
  },
  "actions": {
    "my-logging-action": {
      "logging": {
        "text": "The average request duration for k8s API Server is {{ctx.payload.aggregations.avg_duration}} ms. Threshold is 10 ms."
      }
    }
  },
  "metadata": {
    "xpack": {
      "type": "json"
    }
  }
}
  ```
</details>

**Comment by ChrsMark** [comment]
@gizas I think we already have a good set of Alerts listed in the above comment. All these are fully functional and ready to be used.

 What would the ideal place to "store" those? Maybe somewhere inside the Kubernetes package? Note that due to the long json  bodies it would be better to store those in standalone files. Also I'm not sure if it is allowed by the package spec to store extra folders under the directory structure of a package but I will check it.

**Comment by gizas** [comment]
Really great progress here Chis!!!!

I reviewed them and dont think we miss something important for first iteration.
Some additional ideas:
- A missing agent alert could be a way to identify that cluster has been disconnected? An alert with specific agent name to check rate of data coming. Dont know how much noise will this produce especially if agent restarts. Do we want something like this?
- Also the scenario of no data coming at all from agent can be tackled with latency with system metrics. We can make the assumption that almost always we collect system monitor metrics.
- How about the scenario that we stopped receiving logs? Can we also have something for kubernetes logs?

Now as far as the folder is concerned, I guess `_dev/deploy/alerts` fine for now. We can include an extra file there to track down the versions of ES and package we have tested with? I would suggest to open a new ticket to elastic-package team to start discussing the support of alerts as assets. I would suggest a new story where we would like alerts to be part of a new folder (like dashboards) inside each package.

Reminder that we would like a script to automate the installation of all of this with one run. Ideally the user to choose which to install or not

**Comment by jsoriano** [comment]
> What would the ideal place to "store" those? Maybe somewhere inside the Kubernetes package? Note that due to the long json bodies it would be better to store those in standalone files.
> 
> I tried to put them under the kubernetes package and it is possible to store them under the `_dev/deploy/alerts` subdirectory. See an example draft PR that illustrates the point https://github.com/elastic/integrations/pull/5364/files.
> 
> I also tried to put them under `docs` folder but `elastic-package` complains about this with `item [alerts] is not allowed in folder [/home/chrismark/go/src/github.com/elastic/integrations/build/packages/kubernetes-1.31.2.zip/docs]`.
> 
> For now it would be ok to store them under this subdir and reference those files from our docs maybe? However this would be problematic when it comes to version syncronization etc. @jsoriano any thoughts here on how we could approach this better?

Are these alerts intended for documentation purposes only, or do we want them installed when installing the packages?

If they are only used for documentation, they could be included in the `docs` folder, inside markdown files with more information about how to use them.

If we want them to be installed when installing the package, then we should add them to their own place in the package-spec, and we should have a follow-up task in Kibana to install them when available. This process would start with a new issue in the package-spec repository.

**Comment by gizas** [comment]
The end goal would be to be installed when installing the package!
So we would like to initiate the discussion that alerts are assets of a package

I would suggest for now to include them in doc in order to proceed with our implementation and follow up with the stories as you suggest.

**Comment by jsoriano** [comment]
> The end goal would be to be installed when installing the package!
> So we would like to initiate the discussion that alerts are assets of a package

Perfect, please create an issue in the package-spec repository to discuss about this.

**Comment by ChrsMark** [comment]
Thanks for your feedback here folks! I will file an issue to the package-spec.

As far as the placement into the docs of the package for now, this seems to be problematic because of the `{{ctx.payload}}` reference inside the json request body. It seems that `elastic-package` identifies this a "function" (like the `{ event }` one) and fails to compile the file.
See https://github.com/elastic/integrations/pull/5364/files.
The error I get is:

```console
elastic-package build                  
Build the package
Error: updating files failed: updating readme file apiserver-latency-alert.md failed: rendering Readme failed: parsing README template failed (path: /home/chrismark/go/src/github.com/elastic/integrations/packages/kubernetes/_dev/build/docs/apiserver-latency-alert.md): template: apiserver-latency-alert.md:91: function "ctx" not defined
```

I wonder if we can escape this somehow 🤔 @jsoriano is this met before in any of the packages?

**Comment by jsoriano** [comment]
Files in `_dev/build/docs` are expected to be Go templates that are rendered in `docs`, but I think you can put files directly in `docs`. If you don't need these files to be templates, try to place them in `docs` directly.
If that doesn't work, or you need them to be templates for some reason, you can leave them where they are, but you will need to use raw strings for the code with embedded templates (``{{`{{ctx.payload}}`}}``).

**Comment by ChrsMark** [comment]
Thanks @jsoriano, having them under `docs` directly work fine!

@gizas I have filed a PR to add them inside the package: https://github.com/elastic/integrations/pull/5364
I will proceed with adding them as automation also at https://github.com/elastic/k8s-integration-infra.

**Comment by gizas** [comment]
One last from me here @ChrsMark . I see we now have a full first version of alerts. Can we distinquish those from SLOs?

I mean we can keep current `watchers` folder to include alerts. (Dont know if we should rename to alerts)
Create a new folder called slos. And we can create 1-2 basic slos examples for our Users. SLOs are not so important for now as theys hould come from product requirements but it will be good to showcase the difference, the usage of SLO api and also to guide users how to build them. What do you think? Maybe you can open a new story for this

**Comment by ChrsMark** [comment]
Good point @gizas! One basic limitation that I found in all APIs/methods other than the Watchers one is that those do not provide flexibility to define advanced ES queries with aggregations.

For example if you check the examples at https://github.com/kdelemme/kibana/blob/main/x-pack/plugins/observability/dev_docs/slo.md the SLOs can only be defined on specific fields like `cpu_usage` or `error_rate`. However in a Kubernetes cluster I'm not sure how useful this is since we need to group metrics by Objects like Nodes or Pods so there is no way to make use of the SLO API. 

So in order to implement what we needed to conceptually cover the monitoring/alerting of a k8s cluster we need to go with advanced queries that Watchers allow.

One thing here is that it won't be that great for now to have 12 Watchers/Alerts and then 2 SLOs, to my mind it is better to be consistent.
However I can spend some time to check if there are k8s fields that could be used directly in SLOs like in the examples at https://github.com/kdelemme/kibana/blob/main/x-pack/plugins/observability/dev_docs/slo.md.

**Comment by ChrsMark** [comment]
Hey @gizas I tried to come up with a basic SLO and the only thing that could create looks like the following:

```console
curl -i -k --request POST \
  --url https://elastic:changeme@0.0.0.0:5601/api/observability/slos \
  --header 'Authorization: Basic ZWxhc3RpYzpjaGFuZ2VtZQ==' \
  --header 'Content-Type: application/json' \
  --header 'kbn-xsrf: oui' \
  --data '{
        "name": "Kubernetes Data",
        "description": "My SLO Description",
        "indicator": {
                "type": "sli.kql.custom",
                "params": {
                        "index": "metrics*",
                        "good": "data_stream.dataset: kubernetes.node",
                        "total": "",
                        "filter": ""
                }
        },
        "timeWindow": {
                "duration": "1d",
                "isRolling": true
        },
        "budgetingMethod": "occurrences",
        "objective": {
                "target": 0.001
        }
}'
```

Again, the API looks quite limiting at the moment and for example a numerator query is required while I only wanted to perform a doc count query.
Also the duration cannot be set on minutes. I guess the concept of SLOs implies that the monitoring scopes at a wider timespan and not in minute or second scale.

Having said this I feel that we would need to wait for https://github.com/elastic/kibana/issues/150050 before investing much time in creating SLOs that are not actually useful for the k8s use-case. See what are the plans for the "Custom Metric SLI" which is what we would need most probably -> https://github.com/elastic/kibana/issues/150050#issuecomment-1452030179.

**Comment by ChrsMark** [comment]
A possible demo would look like the following:

1. Install the Watcher for OOMKilled pods:
```console
curl -X PUT "https://elastic:changeme@localhost:9200/_watcher/watch/Pod-Terminated-OOMKilled?pretty" -k -H 'Content-Type: application/json' -d'
{
  "trigger": {
    "schedule": {
      "interval": "1m"
    }
  },
  "input": {
    "search": {
      "request": {
        "search_type": "query_then_fetch",
        "indices": [
          "*"
        ],
        "rest_total_hits_as_int": true,
        "body": {
          "size": 0,
          "query": {
            "bool": {
              "must": [
                {
                  "range": {
                    "@timestamp": {
                      "gte": "now-1m",
                      "lte": "now",
                      "format": "strict_date_optional_time"
                    }
                  }
                },
                {
                  "bool": {
                    "must": [
                      {
                        "query_string": {
                          "query": "data_stream.dataset: kubernetes.state_container",
                          "analyze_wildcard": true
                        }
                      },
                      {
                        "exists": {
                          "field": "kubernetes.container.status.last_terminated_reason"
                        }
                      },
                      {
                        "query_string": {
                          "query": "kubernetes.container.status.last_terminated_reason: OOMKilled",
                          "analyze_wildcard": true
                        }
                      }
                    ],
                    "filter": [],
                    "should": [],
                    "must_not": []
                  }
                }
              ],
              "filter": [],
              "should": [],
              "must_not": []
            }
          },
          "aggs": {
            "pods": {
              "terms": {
                "field": "kubernetes.pod.name",
                "order": {
                  "_key": "asc"
                }
              }
            }
          }
        }
      }
    }
  },
  "condition": {
    "array_compare": {
      "ctx.payload.aggregations.pods.buckets": {
        "path": "doc_count",
        "gte": {
          "value": 1,
          "quantifier": "some"
        }
      }
    }
  },
  "actions": {
    "ping_slack": {
      "foreach": "ctx.payload.aggregations.pods.buckets",
      "max_iterations": 500,
      "webhook": {
        "method": "POST",
        "url": "https://hooks.slack.com/services/T04SW3JHXasdfasdfasdfasdfasdf",
        "body": "{\"channel\": \"#k8s-alerts\", \"username\": \"k8s-cluster-alerting\", \"text\": \"Pod {{ctx.payload.key}} was terminated with status OOMKilled.\"}"
      }
    }
  },
  "metadata": {
    "xpack": {
      "type": "json"
    },
    "name": "Pod Terminated OOMKilled"
  }
}
'
```

2. Then deploy the following Pod that will be OOMKilled:
```yml
apiVersion: v1
kind: Pod
metadata:
  name: memory-demo
  namespace: mem-example
spec:
  containers:
  - name: memory-demo-ctr
    image: polinux/stress
    resources:
      requests:
        memory: "100Mi"
      limits:
        memory: "200Mi"
    command: ["stress"]
    args: ["--vm", "4", "--vm-bytes", "750M", "--vm-hang", "4"]
```

When then watcher will be activated a slack message will be send:
![slackalert](https://user-images.githubusercontent.com/11754898/223616892-78cb3a2c-7532-4dd8-b9b7-5259738e0898.png)


Note that one active slack webhook is required.

cc: @gizas @mlunadia

**Comment by ChrsMark** [comment]
Blogpost was released: https://www.elastic.co/blog/enable-kubernetes-alerting-elastic-observability. Closing this one for now.
